### PR TITLE
fix(core,cli): align context-graph ID validation and bound keystore KDF params

### PR DESCRIPTION
**Pull request by branarakic**
Two small, independent input-bounds gaps. Both are defence-in-depth — neither is reachable as a failure today by an ordinary caller, but both leave a validator weaker than its sibling.

### 1. `core`: two context-graph ID validators disagreed on strictness

`validateContextGraphId` (core) and `isValidContextGraphId` (cli/http-utils) validate the same thing, but only the latter applied a segment-aware check. The character whitelist permits `/` and `.` because URN/DID/URI-shaped IDs need them — which also admits `.` and `..` as whole segments.

That matters because **26 route call sites** reach the core validator via `validateRequiredContextGraphId`, i.e. the weaker of the two.

Fix rejects the *segments* rather than the characters, so ordinary IDs keep working:

| Input | Before | After |
|---|---|---|
| `victim/../secret` | valid | rejected |
| `did:dkg:context-graph:acme.example/v1` | valid | valid |
| `graph/v1..2` | valid | valid |

This aligns core with the check the CLI already had, rather than introducing a new rule.

### 2. `cli`: keystore KDF parameters had a floor but no ceiling

Loading already enforces minimum scrypt parameters — those stop a file advertising a cost cheap enough to attack. There was no maximum, so a file could advertise a cost too expensive to service: scrypt's working set is `128 * N * r`, so `N=2^30, r=8` asks the loader for ~1 TiB before the passphrase is ever checked.

Adds:
- a memory ceiling (512 MiB) on the implied working set
- a parallelism ceiling on `p`
- a power-of-two check on `N` — scrypt requires it, and without the check the value reaches OpenSSL and fails there with an opaque error rather than a diagnosable one

Both ceilings sit **above** the parameters this CLI itself writes (`N=2^18, r=8`, exactly 256 MiB). A regression test pins that, so the ceiling can't later be tightened below our own output.

### Tests

`packages/core` constants — **49 passing**, incl. new cases for rejected segments and for dots that must stay valid.
`packages/cli` keystore — **20 passing**, incl. 3 new bound cases and the self-compatibility guard.

Verified the new tests are real discriminators: on the unmodified base, 3 of them fail. (The `p` case is instructive — it takes ~64s to fail there, because the unbounded value is actually run.)

### CODE REVIEW

File: packages/cli/src/keystore.ts [hunk 1]
```diff
@@ -58,6 +58,23 @@ const MIN_SCRYPT_P = 1;
...
+/**
+ * Upper bounds, the counterpart to the minimums above. The minimums stop a
+ * keystore from advertising a cost so low that the KDF is cheap to attack; the
+ * maximums stop one from advertising a cost so high that simply loading the
+ * file exhausts the process before any passphrase is checked. scrypt's working
+ * set is 128 * N * r bytes: this CLI writes N=2^18, r=8, i.e. exactly 256 MiB,
+ * so the ceiling is set one doubling above that to leave room for keystores
+ * written by other tools at a higher cost, while still bounding the allocation.
+ */
+const MAX_SCRYPT_MEMORY_BYTES = 512 * 1024 * 1024;
```
Comment (otReviewAgent) [inline packages/cli/src/keystore.ts]: 🔴 Bug: The advertised 512 MiB ceiling exceeds the loader's actual 256 MiB limit

**What's wrong**
The added validation declares keystores up to 512 MiB serviceable, but the downstream KDF remains capped at 256 MiB. Accepted parameter sets therefore still fail with an opaque OpenSSL error, defeating both the compatibility allowance and the new diagnostic behavior.

**Example**
A valid external keystore with `n=2**19`, `r=8`, and `p=1` passes the new ceiling check, then fails inside `scryptSync` with “memory limit exceeded” instead of decrypting or receiving the intended up-front diagnostic.

**Suggested direction**
Use one effective memory-budget calculation for both validation and `scryptSync.maxmem`. Include the bounded OpenSSL overhead, or lower the advertised ceiling—but it must still accommodate the CLI's own production parameters.

**For Agents**
Coordinate the limit in `packages/cli/src/keystore.ts` with the `maxmem` passed by `deriveKey`, accounting for OpenSSL overhead while preserving a bounded allocation. Add a boundary check proving the CLI default and the highest advertised supported parameters reach derivation successfully, while the first unsupported cost is rejected before OpenSSL.

Make the KDF ceiling and execution limit one policy

**What's wrong**
The PR adds a second, contradictory resource policy: one limit decides which keystores are accepted and another decides what OpenSSL will execute. That makes the boundary misleading and leaves maintenance spread across constants, inline validation, and derivation internals.

**Example**
Parameters `n=2**19, r=8, p=1` imply exactly 512 MiB, so they pass the new `>` ceiling check. They are then rejected by `scryptSync` because `deriveKey` still supplies a 256 MiB `maxmem`; a direct Node probe returns `ERR_CRYPTO_INVALID_SCRYPT_PARAMS`.

**Suggested direction**
Introduce a single validated scrypt-parameter model or helper and have `deriveKey` consume it, including the same resource limit passed to OpenSSL. This would remove the new inline branching from `decryptKeystore`, eliminate contradictory constants, and allow a number-safe power-of-two check whose domain is explicit.

**For Agents**
Refactor `packages/cli/src/keystore.ts` so one KDF policy/validation helper owns parameter shape, lower bounds, upper bounds, memory calculation, and the execution `maxmem`. Keep the intended accepted and rejected ranges, and prove both sides of the actual execution boundary with focused tests.

File: packages/cli/test/keystore.test.ts [hunk 2]
```diff
@@ -112,6 +112,43 @@ describe('decryptKeystore error handling', () => {
...
+    ks.crypto.kdfparams.p = 1024;
+    await expect(decryptKeystore(ks, PASSPHRASE)).rejects.toThrow(
+      /p too high/,
+    );
+  });
+
+  it('still accepts the parameters this CLI itself writes', async () => {
+    // Guards the ceiling against being set below our own defaults: the CLI
+    // writes N=2**18, r=8, which is exactly 256 MiB of working set.
+    const ks = await encryptKeystore(TEST_KEY, PASSPHRASE);
```
Comment (otReviewAgent) [inline packages/cli/test/keystore.test.ts]: 🔴 Bug: The production-default compatibility test actually uses the reduced test parameter

**What's wrong**
The new test claims to prove that the ceiling accepts parameters written by the CLI, but the global test setup changes what the CLI writes. It therefore gives false confidence that N=2^18 remains accepted.

**Example**
If the memory ceiling were accidentally reduced to 64 MiB, this test would remain green because its keystore costs about 32 MiB, while production keystores costing 256 MiB would be rejected.

**Suggested direction**
Make the test exercise and assert the real production parameters, while keeping expensive derivation isolated from ordinary parallel CI tests.

**For Agents**
In `packages/cli/test/keystore.test.ts`, isolate production-default verification from the suite-wide reduced N. Assert the emitted N explicitly and add a suitably provisioned serial smoke test, or test the parameter-validation boundary separately without performing the expensive derivation.

File: packages/core/src/constants.ts [hunk 3]
```diff
@@ -588,6 +588,15 @@ export function validateContextGraphId(id: string): { valid: boolean; reason?: s
   if (!id || id.length === 0) return { valid: false, reason: 'Context graph ID cannot be empty' };
   if (id.length > 256) return { valid: false, reason: 'Context graph ID exceeds 256 characters' };
   if (!/^[\w:/.@\-]+$/.test(id)) return { valid: false, reason: 'Context graph ID contains disallowed characters (allowed: alphanumeric, _, :, /, ., @, -)' };
+  // The character class above permits both `/` and `.`, because legitimate IDs
+  // are URN/DID/URI-shaped and need them. That combination also admits relative
+  // path segments, which no legitimate ID uses and which do not survive being
+  // joined into a storage namespace or URI with a stable meaning. Reject the
+  // segments themselves rather than the characters, so ordinary IDs like
+  // `did:dkg:context-graph:acme.example/v1` stay valid.
+  if (id.split('/').some((segment) => segment === '.' || segment === '..')) {
```
Comment (otReviewAgent) [inline packages/core/src/constants.ts:601]: 🟡 Issue: Remove the duplicate CLI context-graph validation policy

**What's wrong**
The change aligns the validators by copying the missing rule into core but leaves two full implementations. This fixes today's divergence without removing its cause, so every subsequent policy change still has two canonical-looking places to update.

**Example**
A future change to the allowed characters or 256-character limit in the core validator can leave HTTP routes using the CLI validator with different behavior—the same validator drift this PR is trying to repair.

**Suggested direction**
Make `validateContextGraphId` the sole policy owner. Delete the CLI implementation or reduce it to `typeof id === 'string' && validateContextGraphId(id).valid` so the boolean API remains available without copying validation rules.

**For Agents**
Update `packages/cli/src/daemon/http-utils.ts` and its route call sites to use the core `validateContextGraphId` policy. Preserve a small boolean adapter only if existing imports require it, retaining the runtime string guard where necessary. Verify core and HTTP callers accept and reject the same representative IDs.

File: packages/cli/test/keystore.test.ts [hunk 4]
```diff
@@ -112,6 +112,43 @@ describe('decryptKeystore error handling', () => {
...
     );
   });
+
+  it('rejects keystore declaring a KDF memory cost above the ceiling', async () => {
+    // The existing minimums stop a file advertising a cost cheap enough to
+    // attack. This is the other end: scrypt's working set is 128 * N * r, so
+    // N=2**30 with r=8 asks the loader for ~1 TiB before the passphrase is ever
+    // checked. Rejected up-front with a diagnosable error instead.
+    const ks = await encryptKeystore(TEST_KEY, PASSPHRASE);
+    ks.crypto.kdfparams.n = 2 ** 30;
```
Comment (otReviewAgent) [inline packages/cli/test/keystore.test.ts]: 🟡 Issue: The tests do not verify the security ceilings they introduce

**What's wrong**
The rejection cases use values orders of magnitude above the configured limits. They prove only that some upper bound exists, not that the new 512 MiB and p=16 limits are enforced.

**Example**
Changing `MAX_SCRYPT_MEMORY_BYTES` from 512 MiB to 512 GiB, or `MAX_SCRYPT_P` from 16 to 1000, would leave these tests green despite materially weakening the DoS protection.

**Suggested direction**
Exercise values at and immediately above both declared maxima so changes that weaken or incorrectly exclude the intended boundary fail the suite.

**For Agents**
Add boundary-focused cases in `packages/cli/test/keystore.test.ts`: reject memory immediately above 512 MiB and p=17, and prove 512 MiB and p=16 pass the upper-bound checks. A later-invalid field such as dklen can stop execution before scrypt allocation when testing accepted boundaries.

File: packages/cli/src/keystore.ts [hunk 5]
```diff
@@ -58,9 +58,84 @@ const MIN_SCRYPT_P = 1;
...
+      `Refusing to load keystore: KDF parameters above maximum (p=${p} > ${MAX_SCRYPT_P}). scrypt p too high.`,
+    );
+  }
+}
+
 /** @internal Allow tests to use lighter scrypt params to avoid memory limits */
 export function _setScryptN(n: number) { SCRYPT_N = n; }
 
+/** @internal Exposes the KDF cost policy so tests can pin it without deriving. */
+export function _scryptCostPolicy() {
```
Comment (otReviewAgent) [inline packages/cli/src/keystore.ts]: 🟡 Issue: Model the KDF policy directly instead of exporting a test-only introspection bag

**What's wrong**
The new function is a thin test seam that exposes implementation details as a second policy-shaped API. Because production does not use the returned object, it does not actually establish the claimed single owner and makes future policy changes require coordinating multiple representations.

**Example**
If the production N changes, `encryptKeystore` can use the new `SCRYPT_N` while `_scryptCostPolicy().productionWorkingSetBytes` continues reporting the working set for `2 ** 18`, leaving tests coupled to a stale shadow policy.

**Suggested direction**
Introduce a typed `ScryptPolicy` or focused internal policy module containing the immutable production parameters, derived memory budget, and validation operation. Have production code consume that same model; tests can import the model directly or exercise its public behavior. This removes `_scryptCostPolicy`, the hardcoded duplicate production N, and the artificial module API exposing private helpers.

**For Agents**
Refactor the policy around packages/cli/src/keystore.ts:37-149 into one typed, immutable configuration/model consumed by encryption, derivation, validation, and tests. Preserve the current limits, diagnostics, and production round-trip behavior. If tests need a cheaper N, keep that override distinct from an immutable production default and prove the model cannot report a different production value than encryption uses.

File: packages/cli/test/keystore.test.ts [hunk 6]
```diff
@@ -112,6 +113,103 @@ describe('decryptKeystore error handling', () => {
...
+    // The error text discriminates — a memory rejection would name the cost.
+    const { maxWorkingSetBytes } = SCRYPT_KDF_POLICY;
+    const workingSetBytes = SCRYPT_KDF_POLICY.workingSetBytes.bind(SCRYPT_KDF_POLICY);
+    const atN = maxWorkingSetBytes / (128 * 8);
+    expect(workingSetBytes(atN, 8)).toBe(maxWorkingSetBytes);
+    const ks = await encryptKeystore(TEST_KEY, PASSPHRASE);
+    ks.crypto.kdfparams.n = atN;
+    ks.crypto.kdfparams.r = 8;
+    ks.crypto.kdfparams.dklen = 16;
+    await expect(decryptKeystore(ks, PASSPHRASE)).rejects.toThrow(/dklen/);
```
Comment (otReviewAgent) [inline packages/cli/test/keystore.test.ts:146]: 🔴 Bug: The ceiling test stops before exercising the execution budget

**What's wrong**
The new policy promises that every accepted KDF cost is executable, but the test named as accepting the exact ceiling deliberately forces an earlier dklen validation error. Consequently, it verifies only the custom policy branch and cannot catch the same class of OpenSSL maxmem regression at the newly accepted upper boundary. The surrounding self-referential assertions add confidence in internal arithmetic without validating the external behavior.

**Example**
If executionMaxmemBytes regressed to maxWorkingSetBytes + 1, every added test would still pass: the 256 MiB production round-trip remains below that budget, while the at-ceiling case exits on dklen. A valid keystore at the accepted 512 MiB ceiling could then fail in OpenSSL with ERR_CRYPTO_INVALID_SCRYPT_PARAMS.

**Suggested direction**
Exercise deriveKey/OpenSSL with valid dklen at the maximum accepted cost and assert the intended cap independently. This should fail if the configured maxmem is only arithmetically above the ceiling but still insufficient for OpenSSL overhead.

**For Agents**
In packages/cli/test/keystore.test.ts, preserve the production round-trip but add validation that reaches real scrypt at the maximum accepted working set, ideally using a precomputed valid fixture to require only one expensive derivation or a serialized resource-heavy integration test. Also assert the independently intended 512 MiB cap rather than deriving every expectation from the implementation constant.

**Comment by branarakic** [comment]
Round-2 finding addressed in 0f5976c97, plus a severity correction below.

**🟡 Test-only policy bag** — `_scryptCostPolicy()` is gone. `SCRYPT_KDF_POLICY` is now a frozen, typed model owning the production parameters, both bounds, the working-set arithmetic, the derived execution budget, and the cost validation — and **production consumes it**: encryption, `deriveKey`, and `decryptKeystore` all read this object, as do the tests. The hardcoded duplicate production N is gone with it. Two structural guards replace the drift risk: a module-load assertion that the policy's own production cost passes its own acceptance check, and the round-trip test restores the test override from `SCRYPT_KDF_POLICY.production.n` and asserts the emitted kdfparams equal `SCRYPT_KDF_POLICY.production` exactly.

---

**Severity correction for reviewers on the maxmem fix.** The earlier commit message says "production keystores could not derive." The mechanism is real — `deriveKey`'s hardcoded 256 MiB `maxmem` rejects the module's own N=2^18/r=8 parameters, since OpenSSL enforces `workingSet + overhead <= maxmem` — but the blast radius claim was wrong: `packages/cli/src/keystore.ts` currently has **no production importers**. It isn't referenced by any CLI command, daemon route, or the agent (whose key management is `agent-keystore.ts`, which doesn't use scrypt), and it isn't exported from the package. The bug was real but latent in an orphaned module; nobody hit it. Worth knowing when weighing this PR: the fix makes the module correct at its own advertised parameters, and the module-load assertion prevents the same divergence from recurring if it gains callers.

**Comment by branarakic** [comment]
Round-3 finding addressed in d5cf68022.

**🔴 Ceiling test stopped before the execution budget** — correct: the at-ceiling case exited on dklen before scrypt, so `executionMaxmemBytes` regressing to `maxWorkingSetBytes + 1` would have kept everything green while at-ceiling keystores failed inside OpenSSL.

New case: encrypt cheaply, re-advertise the exact at-ceiling parameters (`n = maxWorkingSetBytes / (128·8)`, `r = 8`), and assert decryption reaches the AES tag check (`Decryption failed`). A wrong key can only exist if scrypt executed the full 512 MiB cost under the execution budget; an insufficient overhead allowance surfaces as `ERR_CRYPTO_INVALID_SCRYPT_PARAMS` instead and fails the test — verified by sabotaging the allowance to 1 byte. One expensive derivation total, serial with the rest of the file.

Also pinned `maxWorkingSetBytes`, `maxP`, `production`, and `minN` as literals so a drifted policy object can't drag the policy-reading assertions along with it.

File: packages/cli/src/daemon/http-utils.ts [hunk 7]
```diff
@@ -1928,28 +1928,19 @@ export function shouldBypassRateLimitForLoopbackTraffic(ip: string, pathname: st
...
-  // segment-aware check is both stricter (still rejects every real
-  // traversal) and tighter (does not produce false-positive 4xx
-  // for valid context-graph IDs that happen to contain `..` inside
-  // a single segment).
-  for (const seg of id.split("/")) {
-    if (seg === "." || seg === "..") return false;
-  }
-  // Allow URNs, DIDs, simple slug-like identifiers, and URIs
-  return /^[\w:/.@\-]+$/.test(id);
+  if (typeof id !== "string") return false;
```
Comment (otReviewAgent) [inline packages/cli/src/daemon/http-utils.ts]: 🟡 Issue: Finish the validator consolidation instead of retaining a second boundary API

**What's wrong**
The new comment says this adapter owns no rules, but its runtime string check is still behavior that the core validator does not own. Keeping both `isValidContextGraphId` and `validateContextGraphId` leaves readers choosing between two almost-identical APIs and keeps context-graph validation coupled to the already 2,000-line HTTP utility module.

**Example**
`/api/context-graph/create` passes a JSON-derived value into this helper. A numeric `id` is rejected by the adapter's runtime type check, while calling the current core validator directly would not provide the same explicit boundary because its contract accepts only `string`. Thus this supposedly rule-free wrapper is still required to understand validation behavior.

**Suggested direction**
Move the runtime type boundary into the canonical core validator, then delete this boolean adapter and use `.valid` at its few remaining call sites.

**For Agents**
Broaden `validateContextGraphId` in `packages/core/src/constants.ts` to accept `unknown` and return the appropriate invalid result for non-strings. Replace the three `isValidContextGraphId` calls with the canonical validator and remove the adapter/imports. Preserve rejection of non-string route inputs and the existing reason-bearing core API.

File: packages/cli/src/keystore.ts [hunk 8]
```diff
@@ -34,43 +34,129 @@ export interface EncryptedKeystore {
...
+  minP: 1,
+  maxP: 16,
+  requiredDklen: 32,
+  minSaltBytes: 16,
+  maxWorkingSetBytes: 512 * 1024 * 1024,
+  executionMaxmemBytes: 512 * 1024 * 1024 + SCRYPT_OVERHEAD_ALLOWANCE_BYTES,
+  workingSetBytes(n: number, r: number): number {
+    return 128 * n * r;
+  },
+  assertCostWithinLimits(n: number, r: number, p: number): void {
```
Comment (otReviewAgent) [inline packages/cli/src/keystore.ts:111]: 🟡 Issue: Malformed numeric KDF parameters are unverified

**What's wrong**
The new validation evidence only covers integer values even though the policy is intended to prevent every unsupported parameter set from reaching OpenSSL. Valid JSON can contain fractional numbers, so this security-sensitive validation contract remains untested for r and p.

**Example**
A JSON keystore can declare `p: 1.5` or `r: 8.5`. These are numbers and satisfy the current minimum/maximum comparisons, so a regression test should verify that the policy rejects them with its own diagnostic instead of allowing a low-level `scryptSync` parameter error.

**Suggested direction**
Exercise fractional and other non-integer numeric inputs for every untrusted scrypt cost field, particularly r and p, and assert policy-level rejection.

**For Agents**
Add table-driven `decryptKeystore` cases for non-integer runtime values in `kdfparams.r` and `kdfparams.p`. Preserve the valid integer boundaries and assert rejection occurs at the policy boundary with a diagnosable keystore error, before derivation.

**Comment by branarakic** [comment]
Round-4 items addressed in ad0fdd290 — and the fractional-parameter finding was a real validation hole, not just a missing test.

**🟡 Non-integer KDF params** — confirmed live: `r=8.5` / `p=1.5` satisfied every `>=`/`<=` bound and reached OpenSSL, which rejects them with the opaque `ERR_OUT_OF_RANGE` this boundary exists to prevent. `assertCostWithinLimits` now requires safe integers for `r` and `p` (`n` was already covered by the power-of-two check), with table-driven tests over fractional, `NaN`, and `Infinity` inputs asserting policy-level rejection.

**🟡 Finish the consolidation** — `validateContextGraphId` now accepts `unknown`, making the runtime type boundary part of the canonical policy; `isValidContextGraphId` is deleted, its three call sites use `.valid` directly, and eleven stale imports are removed across the daemon routes.
